darshan-util/pydarshan/darshan/backend/cffi_backend.py

```diff
@@ -337,49 +339,45 @@ def log_get_generic_record(log, mod_name, dtype='numpy'):
     if mod_name == 'H5D':
         rec['file_rec_id'] = rbuf[0].file_rec_id
 
-    clst = []
-    for i in range(0, len(rbuf[0].counters)):
-        clst.append(rbuf[0].counters[i])
-    rec['counters'] = np.array(clst, dtype=np.int64)
-    cdict = dict(zip(counter_names(mod_name), rec['counters']))
-
-    flst = []
-    for i in range(0, len(rbuf[0].fcounters)):
-        flst.append(rbuf[0].fcounters[i])
-    rec['fcounters'] = np.array(flst, dtype=np.float64)
-    fcdict = dict(zip(fcounter_names(mod_name), rec['fcounters']))
-
-    if dtype == "dict":
-        rec.update({
-            'counters': cdict, 
-            'fcounters': fcdict
-            })
-
-    if dtype == "pandas":
-        rec['id'] = np.uint64(rec['id'])
-        df_c = pd.DataFrame(cdict, index=[0])
-        df_fc = pd.DataFrame(fcdict, index=[0])
-
-        # flip column order (to prepend id and rank)
-        df_c = df_c[df_c.columns[::-1]]
-        df_fc = df_fc[df_fc.columns[::-1]]
-
-        # attach id and rank to counters and fcounters
-        df_c['id'] = rec['id']
-        df_c['rank'] = rec['rank']
-
-        df_fc['id'] = rec['id']
-        df_fc['rank'] = rec['rank']
-
-        # flip column order
-        df_c = df_c[df_c.columns[::-1]]
-        df_fc = df_fc[df_fc.columns[::-1]]
-
-        rec.update({
-            'counters': df_c,
-            'fcounters': df_fc
-            })
-
+    clst = np.zeros(len(rbuf[0].counters), dtype=np.int64)
+    for i in range(clst.size):
+        clst[i] = rbuf[0].counters[i]
+
+    flst = np.zeros(len(rbuf[0].fcounters), dtype=np.float64)
+    for i in range(flst.size):
+        flst[i] = rbuf[0].fcounters[i]
+
+    if c_cols is None:
+        c_cols = counter_names(mod_name)
+    if fc_cols is None:
+        fc_cols = fcounter_names(mod_name)
+
+    if dtype == "numpy":
+        rec['counters'] = clst
+        rec['fcounters'] = flst
+
+    elif dtype == "dict":
+        rec['counters'] = dict(zip(c_cols, clst))
+        rec['fcounters'] = dict(zip(fc_cols, flst))
+
+    elif dtype == "pandas":
+        # prepend id/rank columns
+        new_cols = ["id", "rank"]
+        new_c_cols = new_cols + c_cols
+        new_f_cols = new_cols + fc_cols
+        # prepend the id/rank values
+        id_rank_list = [rec["id"], rec["rank"]]
+        new_clst = np.asarray([id_rank_list + clst.tolist()]).reshape(1, -1)
+        new_flst = np.asarray([id_rank_list + flst.tolist()], dtype=np.float64).reshape(1, -1)
+        # create the dataframes
+        df_c = pd.DataFrame(data=new_clst, columns=new_c_cols)
+        df_fc = pd.DataFrame(data=new_flst, columns=new_f_cols)
+        # correct the data type for the file hash/id
+        df_c['id'] = np.uint64(df_c['id'])
+        df_fc['id'] = np.uint64(df_fc['id'])
```

Comment:
With these changes, generating the dataframes and converting the file hash data type take ~25% and ~57% of the run time, respectively.

If we were willing to allow the data type for the hash to stay a `float64` upon leaving this function (which would break a test), changing it elsewhere (once we have the dataframe containing all records) would likely be higher performance.

Reply:
Performing the same casts to the same id data twice looks a bit awkward maybe. I'll try to take a look later perhaps.

Reply:
I believe the design decision to use integer keys for the hash table in the C code and Python bindings is likely the correct one. Intuitively, avoiding floating point comparisons for hash assignments seems like a huge gain in avoiding collisions/avoidance of ambiguity.

I'm happy to be wrong, though I'd be surprised on that one!

Reply:
I'm not suggesting we keep the hashes as `float64`'s, I'm saying we could *temporarily* keep them as `float64`'s until the entire record dataframe is complete, *then* convert them back to `uint64`'s (because it is likely faster). At a user-level the returned dataframe would be the same as it is right now, the only difference is the individual-record-dataframes returned by `log_get_generic_record()` would have id's of type `float64`.

Reply:
It seems like a bad idea to *ever* represent the hash as an approximate floating point value if it is not an approximate floating point value. There's probably a better way than casting back and forth, even if the danger of an error from the casting chain is low.

Reply:
Okay, I see what you mean. That will make performance improvements here a little more difficult I suppose, but definitely worth keeping them as integers.

Reply:
Getting back to my point about the repeated casting (on top of the swapping data types risk)--this is a bit safer & faster:

```diff
--- a/darshan-util/pydarshan/darshan/backend/cffi_backend.py
+++ b/darshan-util/pydarshan/darshan/backend/cffi_backend.py
@@ -362,6 +362,7 @@ def log_get_generic_record(log, mod_name, c_cols=None, fc_cols=None, dtype='nump
 
     elif dtype == "pandas":
         # prepend id/rank columns
+        rec_id = np.uint64(rec['id'])
         new_cols = ["id", "rank"]
         new_c_cols = new_cols + c_cols
         new_f_cols = new_cols + fc_cols
@@ -373,8 +374,8 @@ def log_get_generic_record(log, mod_name, c_cols=None, fc_cols=None, dtype='nump
         df_c = pd.DataFrame(data=new_clst, columns=new_c_cols)
         df_fc = pd.DataFrame(data=new_flst, columns=new_f_cols)
         # correct the data type for the file hash/id
-        df_c['id'] = np.uint64(df_c['id'])
-        df_fc['id'] = np.uint64(df_fc['id'])
+        df_c['id'] = rec_id
+        df_fc['id'] = rec_id
         # assign the dataframes to the record
         rec['counters'] = df_c
         rec['fcounters'] = df_fc
```

```
       before           after         ratio
     [f602e959]       [7445b823]
     <nawtrey_issue_440_get_generic_record_perf>       <treddy_exp_casting>
-        23.4±2ms       19.1±0.3ms     0.81  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'tests/input/sample-dxt-simple.darshan', 'MPI-IO', False)
-        72.1±3ms         56.5±2ms     0.78  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'tests/input/sample-dxt-simple.darshan', 'POSIX', False)
-      23.9±0.7ms       18.6±0.4ms     0.78  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'tests/input/sample-dxt-simple.darshan', 'MPI-IO', True)
-        72.3±7ms         56.2±1ms     0.78  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'tests/input/sample-dxt-simple.darshan', 'POSIX', True)
-      1.20±0.2ms         904±20μs     0.75  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'examples/example-logs/dxt.darshan', 'STDIO', False)
-      1.31±0.2ms         985±40μs     0.75  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'examples/example-logs/ior_hdf5_example.darshan', 'MPI-IO', False)
-     1.36±0.06ms      1.02±0.01ms     0.75  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'examples/example-logs/dxt.darshan', 'POSIX', True)
-     1.30±0.07ms         974±40μs     0.75  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'examples/example-logs/dxt.darshan', 'MPI-IO', False)
-      1.28±0.2ms         924±20μs     0.72  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'examples/example-logs/ior_hdf5_example.darshan', 'STDIO', True)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

Reply:
Thanks, this was a pretty simple fix that further improved performance. I think the original concerns have been addressed here so I'll resolve this.